rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Alternate `pkgs/unit` structure
+
+- Use a flat directory, e.g. `pkgs.hello` would be in `pkgs/unit/hello`.
+  - Good because it's simpler, both for the user and for the code
+  - Bad because the GitHub web interface only renders the first 1 000 entries (and we have about 10 000 that benefit from this transition, even given the restrictions)
+  - Bad because it makes `git` and file listings slower
+- Use `substring 0 3 name` or `substring 0 4 name`. This was not done because it still leads to directories in `pkgs/unit` containing more than 1 000 entries, leading to the same problems.
+- Use multi-level structure, like a 2-level 2-prefix structure where `hello` is in `pkgs/unit/he/ll/hello`,
+  if packages are less than 4 characters long, we will it out with `-`, e.g. `z` is in `pkgs/unit/z-/--/z`.
+  This is not great because it's more complicated, longer to type and it would improve performance only marginally.
+- Use a dynamic structure where directories are rebalanced when they have too many entries.
+  E.g. `pkgs.foobar` could be in `pkgs/unit/f/foobar` initially.
+  But when there's more than 1 000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/unit/fo/foobar`.
+  This is not great because it's very complex to determine which directory to put a package in, making it bad for contributors.
+
+## Alternate `pkg-fun.nix` filename
+
+- `default.nix`:
+  - Bad because it doesn't have its main benefits here:
+    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
+    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/unit/hell/hello` equally broken regardless of file name.
+  - Not using `default.nix` frees up `default.nix` for a short expression that is actually buildable, e.g. `(import ../.. {}).hello`, although we don't yet have a use case for this that isn't covered by `nix-build ../.. -A $attrname`.
+  - Bad because using `default.nix` would tempt users to invoke `nix-build .` whereas making package functions auto-callable is a known anti-pattern as it duplicates the defaults.
+  - Good because `default.nix` is already a convention most people are used to
+- `package.nix`/`pkg.nix`: Bad, because it makes the migration to a non-function form of overridable packages harder in the future.
```

Comment:
No migration is planned, in fact I don't think this idea of a non-function form has been discussed at all. So I think these names should be reconsidered. Note the discussion around naming in https://github.com/NixOS/rfcs/pull/140#discussion_r1090915765 as well. It will be discussed in the next [NAT meeting](https://github.com/nixpkgs-architecture/#meetings) on monday.

We do plan to make improvements to package declarations in general though, especially for specifying dependencies, which will require changes to the "package function"-style declarations. I think what happens depends on whether this unit directory standard is standardized and opened to third-parties, or whether it's still only internal to Nixpkgs (which is what this RFC describes). If it's standardized we can't really make changes to the structure and semantics of these files anymore. We can still do that if it's Nixpkgs-internal though.